File: app/api/cart_items.py
from flask import Blueprint, request
from flask_login import login_required, current_user
from sqlalchemy.exc import IntegrityError

from app.models import db, CartItem
from app.forms import validation_errors_formatter, CartItemForm

bp = Blueprint("cart_items", __name__, url_prefix="/cart_items")


@bp.route("",  methods=["POST"])
@login_required
def post_cart_item():
    """CREATE"""
    form = CartItemForm()
    form['csrf_token'].data = request.cookies['csrf_token']
    if (form.validate_on_submit()):
        cart_item = CartItem(
            user_id=current_user.id,
            product_id=form.product_id.data,
            quantity=form.quantity.data,
        )
        db.session.add(cart_item)
        db.session.commit()
        return cart_item.to_dict(), 201
    return {'errors': validation_errors_formatter(form, form.errors)}, 400

@bp.route("",  methods=["GET"])
@login_required
def get_cart_items():
    """READ"""
    return CartItem.query.filter(CartItem.user_id == current_user.id)

@bp.route("",  methods=["PUT"])
@login_required
def put_cart_item():
    """UPDATE"""
    cart_item = CartItem.query.filter(CartItem.product_id == form.product_id.data,
                                   CartItem.user_id == current_user.id).first()
    if (not cart_item):
        return "404", 404
    form = CartItem()
    form['csrf_token'].data = request.cookies['csrf_token']
    if form.validate_on_submit():
        cart_item.quantity = form.quantity.data
        db.session.commit()
        return cart_item.to_dict()
    return {'errors': validation_errors_formatter(form, form.errors)}, 400


@bp.route("<int:product_id>",  methods=["DELETE"])
@login_required
def delete_cart_item(product_id):
    """DELETE"""
    try:
        cart_item = CartItem.query.filter(CartItem.product_id == product_id,
                                    CartItem.user_id == current_user.id).first()
        if cart_item:
            db.session.delete(cart_item)
            db.session.commit()
            return {"message": f"Deleted cart item with product id {product_id}"}
        return "404", 404
    except IntegrityError as e:
        return {"errors": {
            "server": "Server failed to delete"
        }}, 500

File: react-app/src/store/cartItems.js
import { csrfFetch } from './csrf';

const POST_CART_ITEM = 'cartItem/POST_CART_ITEM';
const GET_CART_ITEMS = 'cartItem/GET_CART_ITEMS';
const PUT_CART_ITEM = 'cartItem/PUT_CART_ITEM';
const DELETE_CART_ITEM = 'cartItem/DELETE_CART_ITEM';
const CLEAR_CART = 'cartItem/CLEAR_CART';


export const postCartItem = (product_id, quantity) => async dispatch => {
    const response = await csrfFetch('/api/cart_items', {
        method: "POST",
        body: JSON.stringify({ product_id, quantity })
    });
    const cartItem = await response.json();
    dispatch({ type: POST_CART_ITEM, cartItem });
};

export const getCartItems = () => async dispatch => {
    console.log("getCartItems");
    const response = await csrfFetch(`/api/cart_items`);
    const cartItems = await response.json();
    dispatch({ type: GET_CART_ITEMS, cartItems });
};

export const putCartItem = (product_id, quantity) => async dispatch => {
    const response = await csrfFetch(`/api/cart_items`, {
        method: "PUT",
        body: JSON.stringify({ product_id, quantity })
    });
    const cartItem = await response.json();
    dispatch({ type: PUT_CART_ITEM, cartItem });
};

export const deleteCartItem = productId => async dispatch => {
    await csrfFetch(`/api/cart_items/${productId}`, { method: "DELETE" });
    dispatch({ type: DELETE_CART_ITEM, productId });
};

export const clearCart = () => async dispatch => {
    dispatch({ type: CLEAR_CART });
    await csrfFetch(`/api/cart_items`, { method: "DELETE" });
};

export default function cartItemReducer(state = {}, action) {
    const newState = { ...state };
    switch (action.type) {
        case POST_CART_ITEM:
            newState[action.cartItem.product_id] = action.cartItem.quantity;
            return newState;
        case GET_CART_ITEMS:
            for (const cartItem of action.cartItems) {
                newState[cartItem.product_id] = cartItem.quantity;
            }
            return newState;
        case PUT_CART_ITEM:
            newState[action.cartItem.product_id] = action.cartItem.quantity;
            return newState;
        case DELETE_CART_ITEM:
            delete newState[action.cartItem.product_id];
            return newState;
        case CLEAR_CART:
            return {};
        default:
            return state;
    }
};
